src/core/server/elasticsearch/client/client_facade.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { ApiResponse } from '@elastic/elasticsearch';
+import {
+  RequestBody,
+  RequestNDBody,
+  TransportRequestOptions,
+  TransportRequestPromise,
+} from '@elastic/elasticsearch/lib/Transport';
+import * as RequestParams from '@elastic/elasticsearch/api/requestParams';
+
+export interface ClientFacade {
+  bulk<
+    TResponse = Record<string, any>,
+    TRequestBody extends RequestNDBody = Array<Record<string, any>>,
+    TContext = unknown
+  >(
+    params?: RequestParams.Bulk<TRequestBody>,
+    options?: TransportRequestOptions
+  ): TransportRequestPromise<ApiResponse<TResponse, TContext>>;
+
+  asyncSearch: {
+    delete<TResponse = Record<string, any>, TContext = unknown>(
+      params?: RequestParams.AsyncSearchDelete,
+      options?: TransportRequestOptions
+    ): TransportRequestPromise<ApiResponse<TResponse, TContext>>;
+    get<TResponse = Record<string, any>, TContext = unknown>(
+      params?: RequestParams.AsyncSearchGet,
+      options?: TransportRequestOptions
+    ): TransportRequestPromise<ApiResponse<TResponse, TContext>>;
+    submit<
```

Comment:
Another 'detail' regarding this typed/structured replacement of `APICaller` is that it may be a little difficult to migrate the `retryCall` methods used by the SO client

https://github.com/elastic/kibana/blob/bf04235dae35452061cc7ea3d86d96c19a58206c/src/core/server/elasticsearch/retry_call_cluster.ts#L89-L107

Previously `APICaller` was just a method, so wrapping it to retry was rather trivial.  This this new typed interface, I'm unsure what would be the correct solution to achieve the same thing.

Reply:
we can change signature to accept a function:
```js
export function retryCallCluster(fn: () => Promise<T>): Promise<T> {
    return defer(fn())
      .pipe(
        retryWhen((errors) =>
          errors.pipe(
            concatMap((error, i) =>
              iif(
                () => error instanceof legacyElasticsearch.errors.NoConnections,
                timer(1000),
                throwError(error)
              )
            )
          )
        )
      )
      .toPromise();
  };
}
```

Reply:
Yea, that's the 'easier' solution I see. However multiple endpoint/API calls are used in the SO client. wrapping the whole APICaller with that allowed to be sure every call was going to be wrapper with the retry logic. If we wrap each individual methods, we would need to adapt all calls in the SO repository that were based on this retry logic.

Not really seeing another option atm though.